client/scripts/controllers/chain/substrate/staking.ts

```diff
@@ -226,6 +257,93 @@ class SubstrateStaking implements StorageModule {
+        of(api),
+        api.derive.staking.validators(),
+        api.query.staking.currentEra(),
+        from(this._app.chainEvents.rewards()),
+        api.derive.staking.electedInfo()
+      )),
+      flatMap((
+        [api, { validators: currentSet }, era, rewards, electedInfo]:
+        [ApiRx, DeriveStakingValidators, EraIndex, IReward, DeriveStakingElected]
+      ) => {
+        const commission: ICommissionInfo = {};
+
+        electedInfo.info.forEach(({ accountId, validatorPrefs }) => {
+          const key = accountId.toString();
+          commission[key] = (validatorPrefs.commission.unwrap() || new BN(0)).toNumber() / 10_000_000;
+          return commission[key];
+        });
+        // Different runtimes call for different access to stakers: old vs. new
+        const stakersCall = (api.query.staking.stakers)
+          ? api.query.staking.stakers
+          : api.query.staking.erasStakers;
+        // Different staking functions call for different function arguments: old vs. new
+        const stakersCallArgs = (account) => (api.query.staking.stakers)
+          ? account
+          : [era.toString(), account];
+        return combineLatest(
+          stakersCall.multi(currentSet.map((elt) => stakersCallArgs(elt.toString()))),
+          of(currentSet),
+          of(rewards),
+          of(commission)
+        );
+      }),
+      auditTime(100),
+      map(([exposures, accounts, rewards, commissions ] :
+        [Exposure[], AccountId[], IReward, ICommissionInfo ]) => {
+        // coins.div don't support fraction points. multiply and divide the same number will give fraction points.
+        const n = 100000;
+        const validatorRewards: ICommissionInfo = {};
+        let accountsTotalStake = new BN(0);
+
+        accounts.forEach((account, index) => {
+          const key = account.toString();
+          const exposure = exposures[index];
+          let stakeShare = 0;
+          let othersStake: BN = new BN(0);
+
+          exposure.others.forEach((indv) => {
+            othersStake = othersStake.add(indv.value.toBn());
+          });
+          othersStake = othersStake.add(exposure.own.toBn());
+          if (othersStake.gt(new BN(0))) {
+            stakeShare = exposure.own.toBn().muln(n).div(othersStake).toNumber() / n;
+          }
+
+          const totalStake = exposure?.total.toBn() || new BN(0);
+          let totalReward = rewards.validators[key] || 0;
+
+          // if rewards have rewards for chain id, split with every validator equally.
+          if (rewards.validators[this._app.chain.id])
+            totalReward += rewards.validators[this._app.chain.id] / accounts.length;
+
+          accountsTotalStake = accountsTotalStake.add(totalStake);
+
+          const commissionPercent: number = (commissions[key] || 0) / 100;
+
+          let rewardEarn = totalReward * commissionPercent;
+          const othersReward = totalReward - rewardEarn;
+          rewardEarn += stakeShare * othersReward;
```

Comment:
Does this assume you collect the validator's commission? I'm a little unclear exactly what these variables mean/what context they're in.

Reply:
He is taking the reward, first pulling the commission out, then looking at the stake share the validator himself has and adding that fractional reward also to the validators return.

Reply:
What we should also be doing here is calculating the nominator's payment as well, since that APR will help nominators decide if it makes sense to nominate these validators. We're currently only getting the APR that the validator themselves receives.

Reply:
Yea, I can calculate the nominator's payment as well. it's as simple as 

`nominatorPayment = totalReward - rewardEarn; `

rewardEarn = validator's payment, according to [Reward Distribution](https://wiki.polkadot.network/docs/en/learn-staking#reward-distribution)